### PR TITLE
[ioctl] Build hdwallet derive command line into new ioctl

### LINKED ISSUE #3282: [ioctl] build hdwallet derive command line into new ioctl
### What would you like to be added
 - Step 1: create `hdwalletderive.go` and `hdwalletderive_test.go` files in `./ioctl/newcmd/hdwallet`. 
 - Step 2: refactor  command  https://github.com/iotexproject/iotex-core/blob/master/ioctl/cmd/hdwallet/hdwalletderive.go in `hdwalletderive.go`
 - Step 3: add unit test in xx_test.go

#### Notes:
1. Use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to construct the Cobra command.
2. [Output package](https://github.com/iotexproject/iotex-core/blob/master/ioctl/output) is deprecated. `e.g.` `output.PrintError` could be replaced with `errors.Wrap`.
3. Global variables (config.ReadConfig  config.DefaultConfigFile etc.) have been wrapped into client interface. Please use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to access them if needed
4. The logic of new code should be similar to [the old one](https://github.com/iotexproject/iotex-core/blob/master/ioctl/cmd/hdwallet/hdwalletderive.go).
5. In each unit test, the results should be checked whether they are as the expected. 

#### Example:
1. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountinfo.go
2. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountdelete.go
3. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountcreateadd.go

### Why is this needed
Unit tests couldn't be added for the old ioctl command (./ioctl/cmd)

### PR DESCRIPTION
**Pull request by LuckyPigeon**
# Description

Refactor `hdwalletderive` command in new `ioctl`, with the following note.

* Output package is deprecated, replace it with errors package.
* Replace fmt.Println with cmd.Println
* Refactor unit test to cover the modification.

Fixes #3282 

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [x] TestNewHdwalletDeriveCmd

**Test Configuration**:
- Firmware version: Ubuntu 21.04
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by LuckyPigeon** [comment]
This PR will continue after #3283 merged.

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 1]
```diff
@@ -8,46 +8,94 @@ package hdwallet
...
 	if dectxtLen <= 32 {
-		return "", nil, output.NewError(output.ValidationError, "incorrect data", nil)
+		return "", nil, errors.New("incorrect data")
 	}
 
 	mnemonic, hash := dectxt[:dectxtLen-32], dectxt[dectxtLen-32:]
 	if !bytes.Equal(hash, util.HashSHA256(mnemonic)) {
-		return "", nil, output.NewError(output.ValidationError, "password error", nil)
+		return "", nil, errors.New("password error")
 	}
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:100]: split to two function, and move line75~99 into client interface.  consider [#3423 ](https://github.com/iotexproject/iotex-core/pull/3423) together

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 2]
```diff
@@ -59,21 +107,21 @@ func DeriveKey(account, change, index uint32, password string) (string, crypto.P
 	path := hdwallet.MustParseDerivationPath(derivationPath)
 	walletAccount, err := wallet.Derive(path, false)
 	if err != nil {
-		return "", nil, output.NewError(output.InputError, "failed to get account by derive path", err)
+		return "", nil, errors.New("failed to get account by derive path")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:110]: errors.Wrap

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 3]
```diff
@@ -59,21 +107,21 @@ func DeriveKey(account, change, index uint32, password string) (string, crypto.P
...
 	}
 
 	private, err := wallet.PrivateKey(walletAccount)
 	if err != nil {
-		return "", nil, output.NewError(output.InputError, "failed to get private key", err)
+		return "", nil, errors.New("failed to get private key")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:115]: errors.Wrap

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 4]
```diff
@@ -59,21 +107,21 @@ func DeriveKey(account, change, index uint32, password string) (string, crypto.P
...
 	}
 	prvKey, err := crypto.BytesToPrivateKey(ecrypt.FromECDSA(private))
 	if err != nil {
-		return "", nil, output.NewError(output.InputError, "failed to Bytes private key", err)
+		return "", nil, errors.New("failed to Bytes private key")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:119]: errors.Wrap

File: ioctl/client.go [hunk 5]
```diff
@@ -293,6 +296,36 @@ func (c *client) QueryAnalyser(reqData interface{}) (*http.Response, error) {
 	return resp, nil
 }
 
+func (c *client) ExportHdwallet(password string) ([]byte, error) {
```
Comment (huof6829) [inline ioctl/client.go:299]: ExportHdWallet

File: ioctl/client.go [hunk 6]
```diff
@@ -293,6 +296,36 @@ func (c *client) QueryAnalyser(reqData interface{}) (*http.Response, error) {
...
+	// derive key as "m/44'/304'/account'/change/index"
+	hdWalletConfigFile := c.Config().Wallet + "/hdwallet"
+	if !fileutil.FileExists(hdWalletConfigFile) {
+		return nil, errors.New("run 'ioctl hdwallet create' to create your HDWallet first")
+	}
```
Comment (huof6829) [inline ioctl/client.go:304]: a single function for check. it's also used in hdwalletxxx command. note the result message is different in these commands.

File: ioctl/client.go [hunk 7]
```diff
@@ -293,6 +296,36 @@ func (c *client) QueryAnalyser(reqData interface{}) (*http.Response, error) {
 	return resp, nil
 }
 
+func (c *client) ExportHdwallet(password string) ([]byte, error) {
+	// derive key as "m/44'/304'/account'/change/index"
+	hdWalletConfigFile := c.Config().Wallet + "/hdwallet"
```
Comment (huof6829) [inline ioctl/client.go:301]: c.cfg.Wallet

File: ioctl/client.go [hunk 8]
```diff
@@ -293,6 +296,36 @@ func (c *client) QueryAnalyser(reqData interface{}) (*http.Response, error) {
...
+
+	enctxt, err := os.ReadFile(hdWalletConfigFile)
+	if err != nil {
+		return nil, errors.New("failed to read config")
+	}
+
+	enckey := util.HashSHA256([]byte(password))
+	dectxt, err := util.Decrypt(enctxt, enckey)
+	if err != nil {
+		return nil, errors.New("failed to decrypt")
```
Comment (huof6829) [inline ioctl/client.go:314]: errors.Wrap(err,"....")

File: ioctl/client.go [hunk 9]
```diff
@@ -293,6 +296,36 @@ func (c *client) QueryAnalyser(reqData interface{}) (*http.Response, error) {
...
+func (c *client) ExportHdwallet(password string) ([]byte, error) {
+	// derive key as "m/44'/304'/account'/change/index"
+	hdWalletConfigFile := c.Config().Wallet + "/hdwallet"
+	if !fileutil.FileExists(hdWalletConfigFile) {
+		return nil, errors.New("run 'ioctl hdwallet create' to create your HDWallet first")
+	}
+
+	enctxt, err := os.ReadFile(hdWalletConfigFile)
+	if err != nil {
+		return nil, errors.New("failed to read config")
```
Comment (huof6829) [inline ioctl/client.go:308]: errors.Wrap(err,"failed to read hdWallet config")

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 10]
```diff
@@ -7,50 +7,72 @@
...
+			password, err := client.ReadSecret()
+			if err != nil {
+				return errors.New("failed to get password")
+			}
+
+			addr, _, err := DeriveKey(client, account, change, index, password)
+			if err != nil {
+				return err
+			}
+			cmd.Println(fmt.Sprintf("address: %s\n", addr))
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:62]: cmd.Printf

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 11]
```diff
@@ -7,50 +7,72 @@
...
+// NewHdwalletDeriveCmd represents the hdwallet derive command
+func NewHdwalletDeriveCmd(client ioctl.Client) *cobra.Command {
+	use, _ := client.SelectTranslation(_hdwalletDeriveCmdUses)
+	short, _ := client.SelectTranslation(_hdwalletDeriveCmdShorts)
 
-	dectxtLen := len(dectxt)
-	if dectxtLen <= 32 {
-		return "", nil, output.NewError(output.ValidationError, "incorrect data", nil)
-	}
+	cmd := &cobra.Command{
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:40]: return &cobra.Command{

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 12]
```diff
@@ -59,21 +81,21 @@ func DeriveKey(account, change, index uint32, password string) (string, crypto.P
 	path := hdwallet.MustParseDerivationPath(derivationPath)
 	walletAccount, err := wallet.Derive(path, false)
 	if err != nil {
-		return "", nil, output.NewError(output.InputError, "failed to get account by derive path", err)
+		return "", nil, errors.New("failed to get account by derive path")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:84]: errors.Wrap

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 13]
```diff
@@ -59,21 +81,21 @@ func DeriveKey(account, change, index uint32, password string) (string, crypto.P
...
 	}
 
 	private, err := wallet.PrivateKey(walletAccount)
 	if err != nil {
-		return "", nil, output.NewError(output.InputError, "failed to get private key", err)
+		return "", nil, errors.New("failed to get private key")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:89]: errors.Wrap

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 14]
```diff
@@ -59,21 +81,21 @@ func DeriveKey(account, change, index uint32, password string) (string, crypto.P
...
 	}
 	prvKey, err := crypto.BytesToPrivateKey(ecrypt.FromECDSA(private))
 	if err != nil {
-		return "", nil, output.NewError(output.InputError, "failed to Bytes private key", err)
+		return "", nil, errors.New("failed to Bytes private key")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:93]: errors.Wrap

File: ioctl/newcmd/hdwallet/hdwalletderive_test.go [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+package hdwallet
+
+import (
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/iotexproject/iotex-core/ioctl/config"
+	"github.com/iotexproject/iotex-core/ioctl/util"
+	"github.com/iotexproject/iotex-core/test/mock/mock_ioctlclient"
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive_test.go:15]: a single group

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 16]
```diff
@@ -7,50 +7,72 @@
...
+	cmd := &cobra.Command{
+		Use:   use,
+		Short: short,
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.SilenceUsage = true
+			signer := "hdw::" + args[0]
+			account, change, index, err := util.ParseHdwPath(signer)
+			if err != nil {
+				return errors.New("invalid hdwallet key format")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:49]: errors.Wrap

File: ioctl/newcmd/hdwallet/hdwalletderive.go [hunk 17]
```diff
@@ -7,50 +7,72 @@
...
+			}
 
-	mnemonic, hash := dectxt[:dectxtLen-32], dectxt[dectxtLen-32:]
-	if !bytes.Equal(hash, util.HashSHA256(mnemonic)) {
-		return "", nil, output.NewError(output.ValidationError, "password error", nil)
+			cmd.Println("Enter password:")
+			password, err := client.ReadSecret()
+			if err != nil {
+				return errors.New("failed to get password")
```
Comment (huof6829) [inline ioctl/newcmd/hdwallet/hdwalletderive.go:55]: errors.Wrap

**Review by huof6829 — APPROVED** [review]

File: ioctl/newcmd/account/account.go [hunk 18]
```diff
@@ -160,7 +160,7 @@ func PrivateKeyFromSigner(client ioctl.Client, cmd *cobra.Command, signer, passw
 		if err != nil {
 			return nil, errors.Wrap(err, "invalid HDWallet key format")
 		}
```
Comment (Liuhaai) [inline ioctl/newcmd/account/account.go:162]: why is it changed to client?

Comment (LuckyPigeon) [reply ioctl/newcmd/account/account.go:162]: I guess you're talking about `HdwalletMnemonic`.
The main reason is that we don't want to expose `hdwalletConfigFile`, so we add it in `client` struct, which makes all the related function also needs to be inside client interface.

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**